src/rgen/Microsoft.Macios.Generator/DataModel/Method.Generator.cs

```diff
@@ -46,6 +47,50 @@ readonly partial struct Method {
 	public bool UsePlainString
 		=> ExportMethodData.Flags.HasFlag (ObjCBindings.Method.PlainString);
 
+	/// <summary>
+	/// True if the generated code should retain the return value.
+	/// </summary>
+	public bool RetainReturnValue => ExportMethodData.Flags.HasFlag (ObjCBindings.Method.RetainReturnValue);
+
+	/// <summary>
+	/// True if the generated code should release the return value.
+	/// </summary>
+	public bool ReleaseReturnValue => ExportMethodData.Flags.HasFlag (ObjCBindings.Method.ReleaseReturnValue);
+
+	/// <summary>
+	/// True if the method was marked as a factory method.
+	/// </summary>
+	public bool IsFactory => ExportMethodData.Flags.HasFlag (ObjCBindings.Method.Factory);
+
+	/// <summary>
+	/// True if the return type of the method was returned as a proxy object.
+	/// </summary>
+	public bool IsProxy => ExportMethodData.Flags.HasFlag (ObjCBindings.Method.Proxy);
+
+	/// <summary>
+	/// True if the generated method should use a temp return variable.
+	/// </summary>
+	public bool UseTempReturn {
+		get {
+			var byRefParameterCount = Parameters.Count (p => p.ReferenceKind != ReferenceKind.None);
+
+			// based on the configuration flags of the method and the return type we can decide if we need a
+			// temp return type
+			return (Method: this, ByRefParameterCount: byRefParameterCount) switch {
+				// focus first on the flags, since those are manually added and have more precedence
+				{ ByRefParameterCount: > 0 } => true, { Method.ReleaseReturnValue: true } => true, { Method.IsFactory: true } => true, { Method.IsProxy: true } => true, { Method.MarshalNativeExceptions: true, Method.ReturnType.IsVoid: false } => true,
+
+				// focus on the return type
+				{ Method.ReturnType: { IsVoid: false, NeedsStret: true } } => true, { Method.ReturnType: { IsVoid: false, IsWrapped: true } } => true, { Method.ReturnType.IsNativeEnum: true } => true, {
+					Method.ReturnType.SpecialType:
+																																																			 SpecialType.System_Boolean or SpecialType.System_Char or SpecialType.System_Delegate
+				} => true, { Method.ReturnType.IsDelegate: true } => true,
+				// default will be false
+				_ => false
+			};
```

Comment:
IMHO this looks much more complex that it needs to be.

What about just a series of simple if statements?

```cs
// focus first on the flags, since those are manually added and have more precedence
if (byRefParameterCount > 0)
	return true;

if (Method.ReleaseReturnValue)
	return true;

if (Method.IsFactory)
	return true;

if (Method.IsProxy)
	return true;

if (Method.MarshalNativeExceptions)
	return true;

if (!Method.ReturnType.IsVoid)
	return true;

// focus on the return type
...
```

This would be *much* easier to read and reason about.

Reply:
or at least one pattern per line (there's a code formatter messing this up)